optimum/habana/accelerate/utils/transformer_engine.py

```diff
@@ -0,0 +1,137 @@
+            )
+            te_module.weight = module.weight
+
+            if has_bias:
+                te_module.bias = module.bias
+
+            setattr(model, name, te_module)
+        elif isinstance(module, te.Linear) and not to_transformer_engine and _convert_linear:
+            has_bias = module.bias is not None
+            new_module = torch.nn.Linear(
+                module.in_features, module.out_features, bias=has_bias, dtype=module.weight.dtype, device=module.weight.device
+            )
+            new_module.weight.copy_(module.weight)
+            if has_bias:
+                new_module.bias.copy_(module.bias)
+
+            setattr(model, name, new_module)
+        else:
+            _convert_model(module, to_transformer_engine=to_transformer_engine, _convert_linear=_convert_linear)
+
+
+def has_transformer_engine_layers(model):
+    """
+    Returns whether a given model has some `transformer_engine` layer or not.
+    """
+    if not is_fp8_available():
+        raise ImportError("Using `has_transformer_engine_layers` requires transformer_engine to be installed.")
+    for m in model.modules():
+        if isinstance(m, (te.Linear)):
+            return True
+    return False
+
+def convert_model(model):
+    """
+    Converts torch.nn.Linear modules to `transformers_engine` Linear modules.
+    Adapted from: https://github.com/huggingface/accelerate/blob/v0.27.2/src/accelerate/accelerator.py#L1303
+    """
+    if not has_transformer_engine_layers(model):
+        with torch.no_grad():
+            _convert_model(model)
+        model._converted_to_transformer_engine = True
+    return model
+
+def get_fp8_recipe(fp8_recipe_handler):
+    """
+    Creates transformer engine FP8 recipe object.
+    Adapted from: https://github.com/huggingface/accelerate/blob/v0.27.2/src/accelerate/accelerator.py#L1309
+    """
+    kwargs = fp8_recipe_handler.to_dict() if fp8_recipe_handler is not None else {}
+    if "fp8_format" in kwargs:
+        kwargs["fp8_format"] = getattr(te.recipe.Format, kwargs["fp8_format"])
+    fp8_recipe_handler = te.recipe.DelayedScaling(**kwargs)
+    fp8_recipe_handler.backend = "TE"
+    return fp8_recipe_handler
+
+class FP8ContextWrapper:
+    """
+    Helper class for FP8 context related operations.
+    """
+    def __init__(self, ctx, fp8_recipe):
+        self.ctx = ctx
+        self.fp8_ctx = self.create_fp8_context(fp8_recipe)
+
+    def __enter__(self):
+        self.ctx.__enter__()
+        self.fp8_ctx.__enter__()
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.fp8_ctx.__exit__(exc_type, exc_value, exc_traceback)
+        self.ctx.__exit__(exc_type, exc_value, exc_traceback)
+
+    @staticmethod
+    def create_fp8_context(fp8_recipe):
+        return te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe)
+
+    # `_gradient_checkpointing_func` always takes the function to be recomputed as the first argument. The function
```

Comment:
write this as a standard python help string with ''' my comments ''' within function.

Reply:
Done